stl/inc/algorithm

```diff
@@ -4626,6 +4626,155 @@ _FwdIt2 rotate_copy(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Mid, _FwdIt1 _Last, _FwdIt
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+
+            auto _End = _Get_final_iterator_unwrapped<_It>(_First, _STD move(_Last));
+            if (_First == _Mid) {
+                return {_End, _End};
+            }
+
+            if (_Mid == _End) {
+                return {_STD move(_First), _STD move(_End)};
+            }
+
+            if constexpr (random_access_iterator<_It>) {
+                _RANGES _Reverse_common(_First, _Mid);
+                _RANGES _Reverse_common(_Mid, _End);
+                _RANGES _Reverse_common(_First, _End);
+                _First += (_End - _Mid);
+                return {_STD move(_First), _STD move(_End)};
+            } else if constexpr (bidirectional_iterator<_It>) {
+                _RANGES _Reverse_common(_First, _Mid);
+                _RANGES _Reverse_common(_Mid, _End);
+                while (_First != _Mid && _End != _Mid) {
+                    _RANGES iter_swap(_First, --_End);
+                    ++_First;
+                }
+                if (_First == _Mid) {
+                    _RANGES _Reverse_common(_Mid, _End);
+                    return {_STD move(_First), _STD move(_End)};
+                } else {
+                    _RANGES _Reverse_common(_First, _Mid);
+                    return {_STD move(_First), _STD move(_End)};
+                }
+            } else {
+                auto _Next = _Mid;
+                do { // rotate the first cycle
+                    _RANGES iter_swap(_First, _Next);
+                    ++_First;
+                    ++_Next;
+                    if (_First == _Mid) {
+                        _Mid = _Next;
+                    }
+                } while (_Next != _Last);
+
+                auto _Begin = _First;
+
+                while (_Mid != _Last) { // rotate subsequent cycles
+                    _Next = _Mid;
+                    do {
+                        _RANGES iter_swap(_First, _Next);
+                        ++_First;
+                        ++_Next;
+                        if (_First == _Mid) {
+                            _Mid = _Next;
+                        }
+                    } while (_Next != _Last);
+                }
+                return {_STD move(_Begin), _STD move(_End)};
+            }
+        }
+    };
+
+    inline constexpr _Rotate_fn rotate{_Not_quite_object::_Construct_tag{}};
+
+    // ALIAS TEMPLATE rotate_copy_result
+    template <class _In, class _Out>
+    using rotate_copy_result = in_out_result<_In, _Out>;
+
+    // VARIABLE ranges::rotate_copy
+    class _Rotate_copy_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <forward_iterator _It, sentinel_for<_It> _Se, weakly_incrementable _Out>
+            requires indirectly_copyable<_It, _Out>
+        _NODISCARD constexpr rotate_copy_result<_It, _Out> operator()(

```

Comment:
Ditto should not be `_NODISCARD` (2 places)

Reply:
Done